File: js/test_money.js
const assert = require('assert');
// class Dollar {
//     constructor(amount){
//         this.amount = amount;
//     }
//     times(multiplier) {
//         return new Dollar(this.amount * multiplier);
//     }
// }

class Money {
    constructor(amount, currency){
        this.amount = amount;
        this.currency = currency;
    }
    times (multiplier){
        return new Money(this.amount * multiplier, this.currency);
    }
    divide (divisor){
        return new Money(this.amount / divisor, this.currency)
    }
}
let fiver = new Money(5, "USD");
let tenner = new Money(10, "USD");
// 不輸出任何東西是正常的，只有在出錯的時候會顯示
// assert.strictEqual(tenner.amount, 10);
// assert.strictEqual(tenner.currency, "USD");
assert.deepStrictEqual(fiver.times(2), tenner)

let tenEuros = new Money(10, "EUR");
let twentyEuros = new Money(20, "EUR");
// assert.strictEqual(twentyEuros.amount, 20);
// assert.strictEqual(twentyEuros.currency, "EUR");
assert.deepStrictEqual(tenEuros.times(2), twentyEuros)

let originalMoney = new Money(4002, "KRW");
let actualMoneyAfterDivision = originalMoney.divide(4);
let expectedMOneyAfterDivision = new Money(1000.5, "KRW");
// 這個方法可以一次比較物件裡面的所有東西
assert.deepStrictEqual(actualMoneyAfterDivision, expectedMOneyAfterDivision)

class Portfilo {
    constructor(){
        this.moneys = [];
    }
    add(...moneys){
        this.moneys = this.moneys.concat(moneys);
    }
    evaluate(currency) {
        let total = this.moneys.reduce((sum, money) => {
            return sum + money.amount;
        }, 0)
        return new Money(total, currency);
    }
}

let fifteenDollars = new Money(15, "USD");
let portfilo = new Portfilo();
portfilo.add(fiver, tenner);
assert.deepStrictEqual(portfilo.evaluate("USD"), fifteenDollars);

File: py/test_money.py
import unittest

# class Dollar:
#     def __init__(self, amount) -> None:
#         self.amount = amount
#     def times(self, multiplier):
#         return Dollar(self.amount * multiplier)
class Money:
    def __init__(self,amount, currency) -> None:
        self.amount = amount
        self.currency = currency
    def times(self, multiplier):
        return Money(self.amount * multiplier, self.currency)
    def divide(self, divisor):
        return Money(self.amount / divisor, self.currency)
    # 複寫__eq__方法，之後比較簡單很多(建議的做法)
    def __eq__(self, other):
        return self.amount == other.amount and self.currency == other.currency
# 測試類別要是unittest.TestCase的子類別
class TestMoney(unittest.TestCase):
    # 方法要以test作為開頭(不知道是規定或是偏好)
    def testMultiplicationInMoney(self):
        fiver = Money(5, "USD")
        tenner = Money(10, "USD")
        # assertEqual將實際值和期望值做比較
        self.assertEqual(tenner, fiver.times(2))

    def testMultiplicationInEuros(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = Money(20, "EUR")
        # self.assertEqual(20, twentyEuros.amount)
        # self.assertEqual("EUR", twentyEuros.currency)
        self.assertEqual(tenEuros.times(2), twentyEuros)

    def testDivision(self):
        originalMoney = Money(4002, "KRW")
        actualMoneyAfterDivision = originalMoney.divide(4)
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        self.assertEqual(expectedMoneyAfterDivision, actualMoneyAfterDivision)
        # self.assertEqual(expectedMoneyAfterDivision.amount, 
        #                  actualMoneyAfterDivision.amount)
        # self.assertEqual(expectedMoneyAfterDivision.currency,
        #                  actualMoneyAfterDivision.currency)


if __name__ == '__main__':
    unittest.main()